Use [NonNullTypes] to interpret unannotated types in source

jcouv opened the pull request:
In a context where `NonNullTypes` is `true`, an unannotated type is interpreted as non-null. But in a context where it is `false`, then such type is interpreted as null-oblivious.
Annotated types (ie. with `?`) are unaffected.

The main parts of this change
- revive a property on `Symbol` to indicate the presence of the attribute (on symbol or containing symbol, with module symbols terminating the chain)
- add a similar property in `Binder` (generally look for the containing symbol's context, except if a binding flag was set to break cycles)
- the logic to copy modifiers by re-applying transforms now takes the `NonNullTypes` context into account
- the `[NonNullTypes]` attribute is currently recognized syntactically, rather that properly bound. That is a temporary solution to avoid the numerous code cycles that arose. I'll work on breaking those cycles next.

What is not yet done:
- [x] add more verification of symbols
- [ ] `[NonNullTypes]` targeting a return value or a parameter (will be a follow-up PR)
- [ ] add more verification of emitted metadata (for those transformed) (will be a follow-up PR)
- [ ] I intend to unskip some more existing tests related to this attribute (will be a follow-up PR)
- there's one difficult cycle to break in `SourceNamedTypeSymbol.MakeOneDeclaredBases` which I will investigate and fix in a separate PR
- this PR focuses on the `NonNullTypes` attribute in source. The handling of this attribute in metadata will be a follow-up PR

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -803,7 +813,7 @@ private NamespaceOrTypeSymbolWithAnnotations UnwrapAlias(NamespaceOrTypeOrAliasS
             {
                 AliasSymbol discarded;
-                return NamespaceOrTypeSymbolWithAnnotations.Create(Compilation, (NamespaceOrTypeSymbol)UnwrapAlias(symbol.Symbol, out discarded, diagnostics, syntax, basesBeingResolved));
+                return NamespaceOrTypeSymbolWithAnnotations.CreateNonNull(NonNullTypes, (NamespaceOrTypeSymbol)UnwrapAlias(symbol.Symbol, out discarded, diagnostics, syntax, basesBeingResolved));
```
:memo: I will check whether this and the similar change below are covered by test.

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -588,6 +593,11 @@ private bool IsByRefTarget(int slot)
 
+        private void ReportWarningW(SyntaxNode syntax)
+        {
+            ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_ConvertingNullableToNonNullable, syntax);
```
>ReportStaticNullCheckingDiagnostics [](start = 12, length = 35)

:memo: This makes it easy to set a breakpoint to catch all W warnings.

jcouv commented:
@AlekseyTs @cston I've run into some more non-trivial cycles. It got me thinking, does binding really need to care about merging the information from `[NonNullTypes]` and from annotations? What if we kept the two separate during initial binding?

On one side, we'd have the `[NonNullTypes]` information (which mostly comes from attributes, but they are regular well-known attributes if they are not used elsewhere in binding).
On the other side, we'd have the information about "annotated vs. unannotated" (true/false).
To decide if a type is non-null or null-oblivious, you need to put the two pieces of information together (`NonNullTypes` determines the meaning of unannotated).

cston commented:
>                     AttributeTargets.Struct,

`AttributeTargets.All`

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:932 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -588,6 +593,11 @@ private bool IsByRefTarget(int slot)
             return false;
         }
 
+        private void ReportWarningW(SyntaxNode syntax)
```
>ReportWarningW [](start = 21, length = 14)

Consider naming `ReportWWarning`.

cston commented:
>     [NonNullTypes(true)]

Is this necessary?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:1674 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented:
>             s /*T:string*/ .ToString(); // PROTOTYPE(NullableReferenceTypes): Expecting a string! state

Should be `string~`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:1699 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented:
> public struct C<T, NT> where T : S

Type parameter is not used here or in `D<,>`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:2174 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented:
>             compilation.VerifyDiagnostics();

Consider testing `C(nullableField)` and `D(nullableField)` so it clear the `[NonNullTypes]` attributes had an effect.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:2277 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented:
>             comp.VerifyDiagnostics();

There are several tests where there is only `[NonNullTypes(false)]` but no `[NonNullTypes(true)]` and where no diagnostics are expected. Will we know to update these tests if and when the default is switched to `[NonNullTypes(false)]`?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:3489 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

cston commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -2442,17 +2442,17 @@ internal bool HasNullableAttribute(EntityHandle token, out ImmutableArray<bool>
             }
 
-            value = false;
+            value = default;
```
>default [](start = 20, length = 7)

Consider using `false` instead.

cston commented on src/Compilers/Core/Portable/Symbols/Attributes/CommonFieldWellKnownAttributeData.cs:
```diff
@@ -119,5 +119,26 @@ public MarshalPseudoCustomAttributeData MarshallingInformation
+        #region NonNullTypesAttribute
+
+        private bool? _nonNullTypes;
+        public bool? NonNullTypes
```
>public bool? NonNullTypes [](start = 8, length = 25)

Should these `NonNullTypes` properties be on the C# attribute data types only: `FieldWellKnownAttributeData`, etc.?

cston commented on src/Compilers/CSharp/Portable/Binder/BinderFlags.cs:
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// </summary>
     [Flags]
-    internal enum BinderFlags : uint
+    internal enum BinderFlags : ulong
```
>ulong [](start = 32, length = 5)

Is this change needed?

cston replied:
Try `1u << 31`

---
In reply to: [195607839](https://github.com/dotnet/roslyn/pull/27807#discussion_r195607839) [](ancestors = 195607839,195579166)

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -670,6 +681,13 @@ public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
             }
 
+            public override TypeSymbolWithAnnotations AsObliviousReferenceType()
+            {
```
>{ [](start = 12, length = 1)

`Debug.Assert(_isNullable != true);`

cston commented on src/Compilers/CSharp/Portable/Symbols/ParameterSymbol.cs:
```diff
@@ -230,12 +230,13 @@ public object ExplicitDefaultValue
         {
             get
             {
+                Debug.Assert(false); // PROTOTYPE(NullableReferenceTypes): Can we reach this?
```
>// PROTOTYPE(NullableReferenceTypes): Can we reach this? [](start = 37, length = 56)

Perhaps: `F([NotNullTypes]List<object> arg = new List<object>())`

jcouv replied:
I still didn't manage to reach this. I'll keep PROTOTYPE for a follow-up.

---
In reply to: [195588470](https://github.com/dotnet/roslyn/pull/27807#discussion_r195588470) [](ancestors = 195588470)

cston commented on src/Compilers/CSharp/Portable/Symbols/Symbol.cs:
```diff
@@ -832,13 +832,15 @@ public virtual bool HasUnsupportedMetadata
         /// <summary>
-        /// Is module/type/method/field/property/event/parameter definition opted out of nullable warnings. Not valid to call on non-definitions.
+        /// Is module/type/method/field/property/event/parameter definition opted-in/out of treating un-annotated types as non-null.
+        /// This is determined by the presence of the `[NonNullTypes]` attribute and whether the containing module utilizes nullable reference types.
```
>and whether the containing module utilizes nullable reference types [](start = 81, length = 67)

Isn't this also `[NonNullTypes]`?

cston commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/EventWellKnownAttributeData.cs:
```diff
@@ -6,25 +6,25 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     internal sealed class EventWellKnownAttributeData : CommonEventWellKnownAttributeData
     {
-        #region NullableOptOutAttribute
+        #region NonNullTypesAttribute
```
>NonNullTypesAttribute [](start = 16, length = 21)

Seems surprising that some `AttributeData` cases are early, others are not.

cston commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/MethodWellKnownAttributeData.cs:
```diff
@@ -6,25 +6,25 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     internal sealed class MethodWellKnownAttributeData : CommonMethodWellKnownAttributeData
     {
-        #region NullableOptOutAttribute
+        #region NonNullTypesAttribute
```
>NonNullTypesAttribute [](start = 16, length = 21)

This is also on `CommonMethodEarlyWellKnownAttributeData`. Is that correct?

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs:
```diff
@@ -28,6 +28,7 @@ internal sealed class SourceOrdinaryMethodSymbol : SourceMemberMethodSymbol
         private ImmutableArray<ParameterSymbol> _lazyParameters;
         private TypeSymbolWithAnnotations _lazyReturnType;
         private bool _lazyIsVararg;
+        private bool? _nonNullTypesFromAttributes;
```
>_nonNullTypesFromAttributes [](start = 22, length = 27)

`_lazyNonNullTypesFromAttributes`

cston commented:
@jcouv, your suggestion to separate `?` annotations from `[NonNullTypes]` and only combine those when needed sounds useful. There are other places with potential cycles (such as when loading annotated constraints from metadata) where that approach would help.

---
In reply to: [397171192](https://github.com/dotnet/roslyn/pull/27807#issuecomment-397171192) [](ancestors = 397171192)

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1975,6 +1975,10 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
             {
                 return;
             }
+            if (operand.Kind == BoundKind.SuppressNullableWarningExpression)
```
>if [](start = 12, length = 2)

`while`?

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3290,7 +3300,7 @@ private void ReportNullabilityMismatchInArgument(BoundExpression argument, TypeS
         }
 
-        // PROTOTYPE(NullableReferenceTypes): If support for [NullableOptOut] or [NullableOptOutForAssembly]
+        // PROTOTYPE(NullableReferenceTypes): If support for [NonNullTypes] or [NullableOptOutForAssembly]
```
>for [NonNullTypes] or [](start = 57, length = 21)

Does `[NonNullTypes]` affect this method?

jcouv replied:
I removed `GetTypeOrReturnTypeWithAdjustedNullableAnnotations`. It's in a separate commit (number 8).

---
In reply to: [196956897](https://github.com/dotnet/roslyn/pull/27807#discussion_r196956897) [](ancestors = 196956897)

cston commented on src/Compilers/CSharp/Portable/Symbols/Symbol.cs:
```diff
@@ -1324,5 +1327,95 @@ string IFormattable.ToString(string format, IFormatProvider formatProvider)
+                    return result;
+                }
+            }
+            return null;
```
`return SyntaxBasedNonNullTypes(attributes.Attributes);`

cston commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/MethodEarlyWellKnownAttributeData.cs:
```diff
@@ -0,0 +1,16 @@
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
```
Not used.

cston commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs:
```diff
@@ -705,12 +705,13 @@ internal override bool UtilizesNullableReferenceTypes
-                return _module.HasNullableOptOutAttribute(EntityHandle.ModuleDefinition, out optOut) ? optOut : base.NullableOptOut;
+                // The default for PE modules is [NonNullTypes(false)]
+                bool nonNullTypes;
+                return _module.HasNonNullTypesAttribute(EntityHandle.ModuleDefinition, out nonNullTypes) ? nonNullTypes : base.NonNullTypes;
```
>: base.NonNullTypes [](start = 120, length = 19)

Isn't `PEModuleSymbol` the end of the chain?

jcouv replied:
I'll remove `UtilizesNullableReferenceTypes` in a follow-up PR. It wasn't simple enough to include in this PR.

---
In reply to: [196960121](https://github.com/dotnet/roslyn/pull/27807#discussion_r196960121) [](ancestors = 196960121)

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs:
```diff
@@ -587,12 +583,12 @@ public bool IsNullableOptOutForAssembly(AssemblySymbol assembly)
-                var data = GetDecodedWellKnownAttributeData() as ModuleWellKnownAttributeData;
-                return data?.NullableOptOut ?? base.NullableOptOut;
+                // PROTOTYPE(NullableReferenceTypes): temporary solution to avoid cycle
+                return SyntaxBasedNonNullTypes(((SourceAssemblySymbol)this.ContainingAssembly).GetAttributeDeclarations()) ?? base.NonNullTypes;
```
>base.NonNullTypes [](start = 126, length = 17)

Is `SourceModuleSymbol` the end of the chain?

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:
```diff
@@ -69,19 +69,16 @@ internal sealed class SourcePropertySymbol : PropertySymbol, IAttributeTargetSym
             if (isIndexer && !isExplicitInterfaceImplementation)
             {
-                // Evaluate the attributes immediately in case the IndexerNameAttribute has been applied.
+                // Evaluate the early attributes immediately in case the IndexerName attributes were applied.
```
>IndexerName attributes were [](start = 73, length = 27)

Is the change necessary? Can there be multiple attributes?

jcouv commented:
>             compilation.VerifyDiagnostics();

Thanks. I think that spotted a bug in delegate conversions.

---
In reply to: [397410688](https://github.com/dotnet/roslyn/pull/27807#issuecomment-397410688) [](ancestors = 397410688)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:2240 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

jcouv commented:
>             comp.VerifyDiagnostics();

I'll add `[module: NonNullTypes(true)]` to control that.

---
In reply to: [397438743](https://github.com/dotnet/roslyn/pull/27807#issuecomment-397438743) [](ancestors = 397438743)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:3489 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

jcouv commented:
@cston I had to revert the recent commits addressing PR feedback. One change caused regressions in CI (bootstrap build times out) and I'm trying to figure out which. I'll start again from green and try again.

cston commented on src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs:
```diff
@@ -568,13 +569,23 @@ public bool ApplyNullableTransforms(ImmutableArray<bool> transforms, ref int pos
             }
 
-            if (isNullable)
+            //if (!result.IsValueType) // PROTOTYPE(NullableReferenceTypes): this is causing cycle 6 in NullableAttribute_01
```
>// PROTOTYPE(NullableReferenceTypes): this is causing cycle 6 in NullableAttribute_01 [](start = 39, length = 85)

Does this cycle still occur?

jcouv replied:
Yes, there is still a cycle:

```C#
TypeSymbolWithAnnotations.IsValueType.get() Line 333
TypeSymbolWithAnnotations.ApplyNullableTransforms(ImmutableArray<bool> transforms, bool useNonNullTypes, ref int position, out .TypeSymbolWithAnnotations result) Line 572
NamedTypeSymbol.ApplyNullableTransforms(ImmutableArray<bool> transforms, bool useNonNullTypes, ref int position, out .TypeSymbol result) Line 795
TypeSymbolWithAnnotations.ApplyNullableTransforms(ImmutableArray<bool> transforms, bool useNonNullTypes, ref int position, out .TypeSymbolWithAnnotations result) Line 562
Metadata.PE.NullableTypeDecoder.TransformType(.TypeSymbolWithAnnotations metadataType, ImmutableArray<bool> nullableTransformFlags) Line 34
Metadata.PE.NullableTypeDecoder.TransformType(.TypeSymbolWithAnnotations metadataType, System.Reflection.Metadata.EntityHandle targetSymbolToken, .Metadata.PE.PEModuleSymbol containingModule) Line 26
Metadata.PE.PENamedTypeSymbol.DecodeBaseAndInterfaceNullableAndTupleTypes(.NamedTypeSymbol type, System.Reflection.Metadata.EntityHandle handle, .Metadata.PE.PEModuleSymbol moduleSymbol) Line 518
Metadata.PE.PENamedTypeSymbol.MakeDeclaredBaseType() Line 460
Metadata.PE.PENamedTypeSymbol.GetDeclaredBaseType(Roslyn.Utilities.ConsList<Microsoft.CodeAnalysis.CSharp.Symbol> basesBeingResolved) Line 431
Metadata.PE.PENamedTypeSymbol.TypeKind.get() Line 1631
NamedTypeSymbol.IsValueType.get() Line 393
TypeSymbolWithAnnotations.IsValueType.get() Line 333
TypeSymbolWithAnnotations.ApplyNullableTransforms(ImmutableArray<bool> transforms, bool useNonNullTypes, ref int position, out .TypeSymbolWithAnnotations result) Line 572

```

---
In reply to: [198235913](https://github.com/dotnet/roslyn/pull/27807#discussion_r198235913) [](ancestors = 198235913)

cston replied:
Thanks. It should be addressed in #28028.

---
In reply to: [198272037](https://github.com/dotnet/roslyn/pull/27807#discussion_r198272037) [](ancestors = 198272037,198235913)

cston commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/EventWellKnownAttributeData.cs:
```diff
@@ -6,25 +6,5 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
     internal sealed class EventWellKnownAttributeData : CommonEventWellKnownAttributeData
```
>EventWellKnownAttributeData [](start = 26, length = 27)

Please remove this class.

cston commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/MethodEarlyWellKnownAttributeData.cs:
```diff
@@ -0,0 +1,11 @@
+    /// <summary>
+    /// Information decoded from early well-known custom attributes applied on a method.
+    /// </summary>
+    internal sealed class MethodEarlyWellKnownAttributeData : CommonMethodEarlyWellKnownAttributeData
```
>MethodEarlyWellKnownAttributeData [](start = 26, length = 33)

Is this class needed?

cston commented:
>         private SourcePropertySymbol(

Are the changes in this method necessary? If not, consider reverting.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:54 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>             var comp = CreateCompilation(source, options: TestOptions.ReleaseDll, parseOptions: TestOptions.Regular8);

This is the default value. Consider removing `options` argument from new tests.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:870 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>                 Assert.Equal(expectMatch, member.ReturnType.Equals(member.OverriddenMethod.ReturnType, TypeCompareKind.AllIgnoreOptions | TypeCompareKind.CompareNullableModifiersForReferenceTypes));

Why `AllIgnoreOptions` here and below?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:3776 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>     public override string[] P1 {get; set;} // PROTOTYPE(NullableReferenceTypes): I'm expecting a warning here

We should not warn when override is null-oblivious.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:4246 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>             compilation.VerifyDiagnostics();

Should warn for B.M1 and B.M2.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:4818 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>         [Fact(Skip = "[NonNullTypes(true)] is disabled")]

`[NullableOptOutForAssembly]` here and test below.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:23646 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

cston commented:
>         [Fact(Skip = "NonNullTypes does not control warnings")] // PROTOTYPE(NullableReferenceTypes): Update or remove test.

Let's remove these tests that expect the attribute to control warnings. Perhaps in a separate PR.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:24755 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

jcouv commented:
>         [Fact(Skip = "NonNullTypes does not control warnings")] // PROTOTYPE(NullableReferenceTypes): Update or remove test.

I'm expecting to each unskip or remove them, in a follow-up PR.

---
In reply to: [400424239](https://github.com/dotnet/roslyn/pull/27807#issuecomment-400424239) [](ancestors = 400424239)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:24755 in 31429cd. [](commit_id = 31429cd88ba99d331ca9c3b96119f33e26461979, deletion_comment = False)

jcouv commented:
@dotnet/roslyn-compiler for a second review. Thanks

jcouv commented:
>             compilation.VerifyDiagnostics();

@cston From our discussion, I'm disallowing `[NonNullTypes]` on return value and parameters.

---
In reply to: [399260480](https://github.com/dotnet/roslyn/pull/27807#issuecomment-399260480) [](ancestors = 399260480,397410688)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:2240 in e83865f. [](commit_id = e83865f0b49888d86be6e7f1ff810c6e95bea6fb, deletion_comment = False)

jcouv commented:
@333fred for a second review when you get the chance. Thanks

333fred commented on docs/features/nullable-reference-types.md:
```diff
@@ -40,9 +40,18 @@ Unannotated reference types are non-nullable or null-oblivious depending on whet
-    [AttributeUsage(
-        AttributeTargets.All,
-        AllowMultiple = false)]
+    [AttributeUsage(AttributeTargets.Class |
```
Why not assembly?

jcouv replied:
From our discussion, I added a [note](https://github.com/dotnet/roslyn/issues/22152#issuecomment-334331966) to work items (`module:` target is indeed unfamiliar to developers).

333fred commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -996,8 +996,8 @@ private static void CheckRestrictedTypeReceiver(BoundExpression expression, Comp
-                    receiverArgument = CreateConversion(receiver, methodResult.Result.ConversionForArg(0), 
-                                                                     GetTypeOrReturnTypeWithAdjustedNullableAnnotations(receiverParameter).TypeSymbol, 
+                    receiverArgument = CreateConversion(receiver, methodResult.Result.ConversionForArg(0),
+                                                                     receiverParameter.Type.TypeSymbol,
```
>              [](start = 56, length = 13)

Nit: consider reformatting to have consistent indentation with the previous line. Looks like this is a victim of past renames.

jcouv commented on src/Compilers/CSharp/Portable/Symbols/Source/FieldSymbolWithAttributesAndModifiers.cs:
```diff
@@ -336,6 +336,15 @@ internal sealed override bool HasSpecialName
+            get
+            {
+                // PROTOTYPE(NullableReferenceTypes): temporary solution to avoid cycle
+                return SyntaxBasedNonNullTypes(this.AttributeDeclarationSyntaxList) ?? base.NonNullTypes; 
```
@AlekseyTs In case this helps, this is the approach we had previously used for syntax-based detection of NonNullTypes.
The specific `SyntaxBasedNonNullTypes` in this PR was very inefficient, but we could use that model and make the check cheaper (ie. avoid `.ToString()`).
